File: modules/gui/cstl-web-client/src/main/java/org/constellation/gui/StyleEdition.java
/*
 *    Constellation - An open source and standard compliant SDI
 *    http://www.constellation-sdi.org
 *
 *    (C) 2007 - 2012, Geomatys
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 */

package org.constellation.gui;

import com.fasterxml.jackson.databind.ObjectMapper;
import juzu.Action;
import juzu.Path;
import juzu.RequestScoped;
import juzu.Resource;
import juzu.Response;
import juzu.Route;
import juzu.View;
import juzu.impl.request.Request;
import juzu.plugin.ajax.Ajax;
import juzu.template.Template;
import org.constellation.gui.binding.Style;
import org.constellation.gui.service.StyleManager;
import org.geotoolkit.filter.DefaultFilterFactory2;

import javax.inject.Inject;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * StyledLayerDescriptor controller to manage style edition.
 *
 * @author Fabien Bernard (Geomatys).
 * @version 0.9
 * @since 0.9
 */
@RequestScoped
public final class StyleEdition {

    @Inject
    private StyleManager api;

    @Inject
    @Path("style/style.gtmpl")
    Template style;

    @Inject
    @Path("style/rule_manual.gtmpl")
    Template rule_manual;

    @Inject
    @Path("style/rule_ai.gtmpl")
    Template rule_ai;

    @Inject
    @Path("style/rule_auv.gtmpl")
    Template rule_auv;

    @Inject
    @Path("style/symbol_point.gtmpl")
    Template symbol_point;

    @Inject
    @Path("style/symbol_line.gtmpl")
    Template symbol_line;

    @Inject
    @Path("style/symbol_polygon.gtmpl")
    Template symbol_polygon;

    @Inject
    @Path("style/symbol_text.gtmpl")
    Template symbol_text;

    @Inject
    @Path("style/symbol_raster.gtmpl")
    Template symbol_raster;

    @View
    @Route("edit/style")
    public Response style(final String providerId, final String styleName, final String load) throws IOException {
        // Prepare input parameters.
        final Map<String, Object> parameters = new HashMap<String, Object>(0);
        parameters.put("providerId", providerId);
        parameters.put("styleName",  styleName);

        // Put the complete style JSON as input parameter if needed.
        if (Boolean.parseBoolean(load)) {
            parameters.put("style", api.getStyleJSON(providerId, styleName));
        } else {
            parameters.put("style", null);
        }

        // Go to view.
        return style.ok(parameters).withMimeType("text/html");
    }

    @View
    @Route("edit/rule")
    public Response rule(final String providerId, final String styleName, final String method, final String ruleIndex) {
        // Prepare input parameters.
        final Map<String, Object> parameters = new HashMap<String, Object>(0);
        parameters.put("providerId", providerId);
        parameters.put("styleName",  styleName);

        // Go to view according the specified edition method.
        if ("m".equals(method)) {
            parameters.put("ruleIndex", ruleIndex);
            return rule_manual.ok(parameters).withMimeType("text/html");
        } else if ("ai".equals(method)) {
            return rule_ai.ok(parameters).withMimeType("text/html");
        } else if ("auv".equals(method)) {
            return rule_auv.ok(parameters).withMimeType("text/html");
        }

        // Not found.
        return Response.notFound();
    }

    @View
    @Route("edit/symbol")
    public Response symbol(final String providerId, final String styleName, final String ruleIndex, final String type, final String symbolIndex) {
        // Prepare input parameters.
        final Map<String, Object> parameters = new HashMap<String, Object>(0);
        parameters.put("providerId",  providerId);
        parameters.put("styleName",   styleName);
        parameters.put("ruleIndex",   ruleIndex);
        parameters.put("symbolIndex", symbolIndex);

        // Go to view according the specified symbolizer type.
        if ("point".equals(type)) {
            return symbol_point.ok(parameters).withMimeType("text/html");
        } else if ("line".equals(type)) {
            return symbol_line.ok(parameters).withMimeType("text/html");
        } else if ("polygon".equals(type)) {
            return symbol_polygon.ok(parameters).withMimeType("text/html");
        } else if ("text".equals(type)) {
            return symbol_text.ok(parameters).withMimeType("text/html");
        } else if ("raster".equals(type)) {
            return symbol_raster.ok(parameters).withMimeType("text/html");
        }

        // Not found.
        return Response.notFound();
    }

    @Ajax
    @Resource
    @Route("edit/update")
    public Response update(final String providerId, final String styleName) {
        try {
            // Get style json body.
            final String json = Request.getCurrent().getParameters().get("style").getValue();

            // Update the style.
            api.updateStyleJSON(providerId, styleName, json);

            // Prepare input parameters.
            final Map<String, Object> parameters = new HashMap<String, Object>(0);
            parameters.put("providerId", providerId);
            parameters.put("styleName",  styleName);

            // Return status
            return Response.status(200);
        } catch (Exception ex) {
            return Response.error(ex);
        }
    }
}

File: modules/gui/style/src/main/resources/org/constellation/gui/assets/js/style_edition.js
/*
 *    Constellation - An open source and standard compliant SDI
 *    http://www.constellation-sdi.org
 *
 *    (C) 2007 - 2012, Geomatys
 *
 *    This library is free software; you can redistribute it and/or
 *    modify it under the terms of the GNU Lesser General Public
 *    License as published by the Free Software Foundation; either
 *    version 3 of the License, or (at your option) any later version.
 *
 *    This library is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *    Lesser General Public License for more details.
 */

/**
 * @author Fabien Bernard (Geomatys).
 * @version 0.9
 * @since 0.9
 */

/**
 * Constellation namespace and utility functions.
 *
 * @type {Object}
 */
var CSTL = {

    i18n: function(key) {
        return CSTL.msg[OpenLayers.Lang.getCode()][key];
    }
};

/**
 * Constellation message bundle registry.
 *
 * @type {Object}
 */
CSTL.msg = {};
CSTL.msg["fr"] = {
    "comparison":     "Comparaison",
    "delete":         "Supprimer",
    "edit":           "Editer",
    "logical":        "Logique",
    "no.description": "Aucune description disponible."
};
CSTL.msg["en"] = {
    "comparison":     "Comparison",
    "delete":         "Delete",
    "edit":           "Edit",
    "logical":        "Logical",
    "no.description": "No description available."
};

/**
 * Constellation UI Style namespace.
 */
CSTL.STYLE = {

    // Default style rule.
    DEFAULT_RULE: {
        name: "Change me!",
        title: "",
        description: "",
        symbolizers: [],
        minScale: 0,
        maxScale: 1.7976931348623157E308,
        filter: null
    },

    // Default symbolizer by type.
    DEFAULT_SYMBOL: {
        'point': {
            '@symbol': 'point',
            graphic: {
                size:     10,
                opacity:  1.0,
                rotation: 0,
                mark: {
                    geometry: 'circle',
                    stroke: {
                        color:   "#000000",
                        opacity: 1.0,
                        width:   2,
                        dashed:  false
                    },
                    fill: {
                        color:   "#ffffff",
                        opacity: 1.0
                    }
                }
            }
        },
        'line': {
            '@symbol': 'line',
            stroke: {
                color:   "#000000",
                opacity: 1.0,
                width:   2,
                dashed:  false
            }
        },
        'polygon': {
            '@symbol': 'polygon',
            stroke: {
                color:   "#000000",
                opacity: 1.0,
                width:   2,
                dashed:  false
            },
            fill: {
                color:     "#ffffff",
                opacity:   1.0
            }
        },
        'text': {
            '@symbol': 'text',
            label: "",
            font: {
                size: 12,
                bold: false,
                italic: false
            },
            fill: {
                color:   "#000000",
                opacity: 1.0
            }
        },
        'raster': {
            '@symbol': 'raster',
            opacity: 1.0,
            colorMap: null,
            channelSelection: {
                greyChannel: null,
                rgbChannels: null
            }
        }
    }
};

CSTL.STYLE.Workflow = {

    current: CSTL.STYLE.StyleEdition,

    validate: function() {
        this.current.validate();
        this.updateOverview();
    },

    updateOverview: function() {
        map.layers[0].params['SLD_BODY'] = $.toJSON(CSTL.STYLE.StyleEdition.getCurrent());
        map.layers[0].redraw(true);
    },

    previous: function() {
        this.current.previous();
        this.updateOverview();
    }
};

/**
 * Global style edition controller.
 */
CSTL.STYLE.StyleEdition = {

    $panel: $('[data-edition="style"]'),
    style: null,
    dataDescription: null,

    init: function() {
        this.$panel = $('[data-edition="style"]');

        // Init all panels.
        CSTL.STYLE.RuleEdition['m'].init();
        CSTL.STYLE.RuleEdition['auv'].init();
        CSTL.STYLE.RuleEdition['ai'].init();
        CSTL.STYLE.SymbolEdition['point'].init();
        CSTL.STYLE.SymbolEdition['line'].init();
        CSTL.STYLE.SymbolEdition['polygon'].init();
        CSTL.STYLE.SymbolEdition['text'].init();
        CSTL.STYLE.SymbolEdition['raster'].init();

        // Opacity sliders.
        $("[name$='.opacity']").slider({min:0,max:100,step:1});

        // Rotation sliders.
        $("[name$='.rotation']").slider({min:0,max:360,step:1});

        // Colorpickers.
        $("[name$='.color']").colorpicker({format:'hex'}).on('changeColor', function(e) {
            $(this).siblings(".color-overview").css("background-color", e.color.toHex());
        });

        // Button groups.
        $('[data-toggle="buttons-radio"] [data-toggle="buttons-checkbox"]').button();
    },

    start: function(style, dataExtent) {
        this.style = style;

        // Draw rule list.
        this.drawRules();

        // Display panel.
        this.$panel.show();

        // Zoom to data extent.
        if (dataExtent) {
            map.zoomToExtent(new OpenLayers.Bounds(dataExtent));
        } else {
            map.zoomToExtent(new OpenLayers.Bounds(-180, -90, 180, 90));
        }

        // Display overview.
        CSTL.STYLE.Workflow.updateOverview();
    },

    drawRules: function() {
        var $rules = this.$panel.find("#rules");
        $rules.empty();

        // For each rule
        for (var i = 0; i < this.style.rules.length; i++) {
            var rule = this.style.rules[i];

            // Row
            var $row = $("<div />", {
                "class": "item rule",
                "data-content": rule.description || CSTL.i18n("no.description")
            }).appendTo($rules);

            // Title
            $("<span />", {
                "class": "title",
                "text": rule.title || rule.name
            }).appendTo($row);

            // Remove link
            $("<a />", {
                "text": CSTL.i18n("delete"),
                "href": "#"
            }).click(function() {
                var $parent = $(this).parent();
                CSTL.STYLE.StyleEdition.removeRule($parent.index());
                $parent.popover("destroy").remove();
                return false;
            }).appendTo($row);

            // Separator
            $("<span />", {
                "class": "vertical-separator"
            }).appendTo($row);

            // Edit link
            $("<a />", {
                "text": CSTL.i18n("edit"),
                "href": "#"
            }).click(function() {
                var $parent = $(this).parent();
                CSTL.STYLE.StyleEdition.editRule($parent.index());
                return false;
            }).appendTo($row);

            // Add popover
            $row.popover({placement:"bottom", trigger:"hover", title:"Description"});
        }
    },

    show: function() {
        this.$panel.show();
        CSTL.STYLE.Workflow.current = this;
    },

    hide: function() {
        this.$panel.hide();
    },

    removeRule:function(index) {
        this.style.rules.splice(index, 1);
        CSTL.STYLE.Workflow.updateOverview();
    },

    removeAllRules:function() {
        this.style.rules = [];
        this.drawRules();
        CSTL.STYLE.Workflow.updateOverview();
    },

    newRule: function(method) {
        this.hide();
        if (method === 'm') {
            CSTL.STYLE.RuleEdition['m'].start(CSTL.STYLE.DEFAULT_RULE, null);
        } else {
            CSTL.STYLE.RuleEdition[method].start();
        }
    },

    editRule: function(index) {
        this.hide();
        CSTL.STYLE.RuleEdition['m'].start(this.style.rules[index], index);
    },

    getCurrent: function() {
        var clone = $.extend(true, {}, this.style);

        // Apply rule changes if any.
        CSTL.STYLE.RuleEdition['m'].updateStyle(clone);
        CSTL.STYLE.RuleEdition['auv'].updateStyle(clone);
        CSTL.STYLE.RuleEdition['ai'].updateStyle(clone);

        return clone;
    },

    previous: function() {

    },

    validate: function() {
        var form = $('#updateForm');
        form.find('input[type=hidden]').val($.toJSON(this.style));
        form.find('input[type=submit]').trigger('click');
    }
};